gpt_index/readers/github_readers/github_api_client.py

```diff
@@ -0,0 +1,370 @@
+        self,
+        github_token: Optional[str] = None,
+        base_url: str = DEFAULT_BASE_URL,
+        api_version: str = DEFAULT_API_VERSION,
+    ) -> None:
+        """
+        Initialize the GithubClient.
+
+        Args:
+            - github_token (str): Github token for authentication. If not provided, the client will try to get it from the GITHUB_TOKEN environment variable.
+            - base_url (str): Base URL for the Github API (defaults to "https://api.github.com").
+            - api_version (str): Github API version (defaults to "2022-11-28").
+
+        Raises:
+            ValueError: If no Github token is provided.
+        """
+        if github_token is None:
+            self.github_token = os.getenv("GITHUB_TOKEN")
+            if self.github_token is None:
+                raise ValueError(
+                    "Please provide a Github token. "
+                    + "You can do so by passing it as an argument to the GithubReader,"
+                    + "or by setting the GITHUB_TOKEN environment variable."
+                )
+        else:
+            self.github_token = github_token
+        self.base_url = base_url
+        self.api_version = api_version
+
+        self.__endpoints = {
+            "getTree": "/repos/{owner}/{repo}/git/trees/{tree_sha}",
+            "getBranch": "/repos/{owner}/{repo}/branches/{branch}",
+            "getBlob": "/repos/{owner}/{repo}/git/blobs/{file_sha}",
+            "getCommit": "/repos/{owner}/{repo}/commits/{commit_sha}",
+        }
+
+        self.__headers = {
+            "Accept": "application/vnd.github+json",
+            "Authorization": f"Bearer {self.github_token}",
+            "X-GitHub-Api-Version": f"{self.api_version}",
+        }
+
+    def get_all_endpoints(self) -> Dict[str, str]:
+        """Get all available endpoints."""
+        return {**self.__endpoints}
+
+    async def request(
+        self,
+        endpoint: str,
+        method: str,
+        headers: Dict[str, Any] = {},
+        **kwargs: Any,
+    ) -> Any:
+        """
+        Make an API request to the Github API.
+
+        This method is used for making API requests to the Github API.
+        It is used internally by the other methods in the client.
+
+        Args:
+            - `endpoint (str)`: Name of the endpoint to make the request to.
+            - `method (str)`: HTTP method to use for the request.
+            - `headers (dict)`: HTTP headers to include in the request.
+            - `**kwargs`: Keyword arguments to pass to the endpoint URL.
+
+        Returns:
+            - `response (httpx.Response)`: Response from the API request.
+
+        Raises:
+            - ImportError: If the `httpx` library is not installed.
+            - httpx.HTTPError: If the API request fails.
+
+        Examples:
+            >>> response = client.request("getTree", "GET", owner="owner", repo="repo", tree_sha="tree_sha")
+        """
+        try:
+            import httpx
```

Comment:
out of curiosity, why do we need httpx for this vs. the standard requests module?

Reply:
Thank you @jerryjliu for taking your time and reviewing.
I initially started with requests module but I found it quite slow especially with big repositories. I was also testing the `GithubRepositoryReader`'s load_data function with multiple repositories at once. I found it that using async request helps with retrieving blobs lot faster because it allows me to make `n` blobs request at once, if there are 100 blobs and `n = 2` then it only takes 50 blobs time request instead of 100 blobs. However I have another branch that uses requests. I haven't pushed it to local.

All the dataclass are just for convenience and add typing to GitHub API endpoints. `BufferedAsyncIterator` for utilizing concurrent requests and lazy-loading-like behavior.

if we take out the async aspect of this PR from equation. We won't be needing most of the utils. I specifically opened this one first to  get your feedback.

Reply:
I see, thanks for the explanation. This makes sense; let me make some time to test this out

Reply:
you can use this branch to test synchronous version [ahmetkca/github-reader-synchronous](https://github.com/ahmetkca/gpt_index/tree/github-reader-synchronous) as well